servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/OptionalSslChannelSingle.java

```diff
@@ -0,0 +1,117 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.SingleSource;
+import io.servicetalk.transport.netty.internal.ChannelInitializer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+final class OptionalSslChannelSingle extends ChannelInitSingle<Boolean> {
+
+    OptionalSslChannelSingle(final Channel channel, final ChannelInitializer channelInitializer) {
+        super(channel, channelInitializer);
+    }
+
+    @Override
+    protected ChannelHandler newChannelHandler(final Subscriber<? super Boolean> subscriber) {
+        return new OptionalSslHandler(subscriber);
+    }
+
+    static final class OptionalSslHandler extends ByteToMessageDecoder {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(OptionalSslHandler.class);
+
+        /**
+         * the length of the ssl record header (in bytes)
+         */
+        private static final int SSL_RECORD_HEADER_LENGTH = 5;
+
+        @Nullable
+        SingleSource.Subscriber<? super Boolean> subscriber;
+
+        OptionalSslHandler(final SingleSource.Subscriber<? super Boolean> subscriber) {
+            this.subscriber = subscriber;
+        }
+
+        @Override
+        public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+            if (ctx.channel().isActive()) {
+                ctx.read(); // we need to force a read to detect SSL yes/no
+            }
+            super.handlerAdded(ctx);
+        }
+
+        @Override
+        public void channelActive(final ChannelHandlerContext ctx) throws Exception {
+            ctx.read(); // we need to force a read to detect SSL yes/no
+            ctx.fireChannelActive();
+        }
+
+        @Override
+        protected void decode(final ChannelHandlerContext ctx, final ByteBuf in, final List<Object> out) {
+            if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH || subscriber == null) {
+                return;
+            }
+            boolean isEncrypted = SslHandler.isEncrypted(in);
+            LOGGER.debug("{} Detected TLS for this connection: {}", ctx.channel(), isEncrypted);
+            final SingleSource.Subscriber<? super Boolean> subscriberCopy = subscriber;
+            subscriber = null;
+            subscriberCopy.onSuccess(isEncrypted);
```

Comment:
Can you try doing `ctx.pipeline().remove(this);` right after this line instead of handling `userEventTriggered`?

Reply:
🤯 it works now -- I can swear it didn't work before but I clearly missed something when developing.

Reply:
This pattern relies on the `Subscriber` to synchronously triggering all pipeline modifications. While this is true for the code as it stands now, it relies on a coupling between the execution models of the netty pipeline and the RS implementation and the consumption of the signal itself and there is no guarantee that the downstream consumers of the `Single` will do their pipeline transformations synchronously forever. That is why I kind of liked having the event that acted as a bridge between the two threading models. If we want to rely on the implicit coupling we should make it it very clear (maybe via asserts?) that the pipeline was modified after calling the subscriber and before we remove this handler so we know someone will catch the bytes.

Reply:
I'll see if I can add some asserts (i.e. assert that the pipeline is not empty before removing the handler) -- the event approach had the problem that there wasn't a central spot where the event got fired, but I had to add it "ad-hoc" in all kinds of places depending on the code path.

Reply:
Bridging the two models is always going to be a bit messy. The cleanest way to do this is to put the initialization all in channel handlers since then you're only dealing with a single execution model but that ship has sailed based on the existing `*ChannelSingle` implementations.